### PR TITLE
[windows] properly compute programdata rather than hard-code to c:

### PR DESCRIPTION
**Pull request by derekwbrown**
### What does this PR do?

Properly computes the location of the `programdata` directory, instead of
assuming it's on `c:`

### Motivation

customer feedback

### Additional Notes

For completeness, will require forthcoming PRs for process & trace

### CODE REVIEW

**Review by trishankatdatadog — CHANGES_REQUESTED** [review]
Thanks for the PR, Derek!

It LGTM for the most part. One issue I see is that if getting the Program Data directory fails for whatever, we fallback to the default value, and the user might run into problems anyway. But I don't see a good way around this right now.

File: cmd/agent/app/integrations_windows.go [hunk 1]
```diff
@@ -24,6 +26,12 @@ var (
 	tufPipCachePath      = filepath.Join("c:\\", "ProgramData", "Datadog", "repositories", "cache")
 )
 
+func init() {
```
[the remark this answers is not in the document]
Comment (derekwbrown) [reply cmd/agent/app/integrations_windows.go]: init() gets called on initialization, prior to `main`

File: pkg/util/winutil/shutil.go [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+}
+
+// GetProgramDataDir() returns the current programdatadir, usually
+// c:\programdata
+func GetProgramDataDir() (path string, err error) {
+	var retstr uintptr
+	err = SHGetKnownFolderPath(&FOLDERID_ProgramData, 0, 0, &retstr)
+	if err == nil {
+		// convert the string
+		defer CoTaskMemFree(retstr)
```
[the remark this answers is not in the document]
Comment (derekwbrown) [reply pkg/util/winutil/shutil.go]: well, yess it's possible that it's non-nil, but if the error is set then retstr can't be counted on to be valid to do anything with, including free.

File: cmd/agent/app/integrations_windows.go [hunk 3]
```diff
@@ -11,6 +11,8 @@ package app
 import (
 	"os"
 	"path/filepath"
+
+	"github.com/DataDog/datadog-agent/pkg/util/winutil"
 )
 
 const (
```
[the remark this answers is not in the document]
Comment (truthbk) [reply cmd/agent/app/integrations_windows.go]: not really, no.

File: pkg/util/winutil/shutil.go [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+var (
+	modShell32               = syscall.NewLazyDLL("Shell32.dll")
+	modOle32                 = syscall.NewLazyDLL("Ole32.dll")
+	procSHGetKnownFolderPath = modShell32.NewProc("SHGetKnownFolderPath")
+	procCoTaskMemFree        = modOle32.NewProc("CoTaskMemFree")
+)
+
+// SHGetKnownFolderPath syscall to windows native SHGetKnownFOlderPath
+func SHGetKnownFolderPath(rfid *GUID, dwFlags uint32, hToken syscall.Handle, pszPath *uintptr) (retval error) {
```
Comment (truthbk) [inline pkg/util/winutil/shutil.go]: We probably don't need to export this or `CoTaskMemFree` below beyond the `winutil` module, so I'd probably keep these as `sHGetKnownFolderPath` and `coTaskMemFree`.

File: pkg/util/winutil/shutil.go [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+// GetProgramDataDir() returns the current programdatadir, usually
+// c:\programdata
+func GetProgramDataDir() (path string, err error) {
+	var retstr *C.char
+	var retstrptr = uintptr(unsafe.Pointer(retstr))
+	err = SHGetKnownFolderPath(&FOLDERID_ProgramData, 0, 0, &retstrptr)
+	if err == nil {
+		// convert the string
+		defer CoTaskMemFree(retstrptr)
+		// go vet: "possible misuse of unsafe.Pointer"
```
Comment (truthbk) [inline pkg/util/winutil/shutil.go]: can you provide some context?

**Review by truthbk — COMMENTED** [review]
This looks great for the most part, I'm not 100% sure about where we should initialize all of this stuff: `init()` blocks vs explicit initalization some other way.... The syscall look good, pointer safety looks good, memory management looks good too.

File: pkg/util/winutil/shutil.go (hunk 5 above)
Comment (derekwbrown) [reply pkg/util/winutil/shutil.go]: Added to the comment.
IMO the original code would be better (b/c it doesn't require `cgo`), but I couldn't get it past CI.

File: pkg/util/winutil/shutil.go (hunk 4 above)
Comment (derekwbrown) [reply pkg/util/winutil/shutil.go]: It's being used in parallel PRs for `trace` & `process`. I think it has to be exported.

File: cmd/agent/common/common_windows.go [hunk 6]
```diff
@@ -45,6 +46,15 @@ const (
 	DefaultDCALogFile = "c:\\programdata\\datadog\\logs\\cluster-agent.log"
 )
 
+func init() {
```
[the remark this answers is not in the document]
Comment (derekwbrown) [reply cmd/agent/common/common_windows.go]: This was the minimum code change to make it work.  We could do some explicity initialization, but
1) this "should" never fail
2) in the event that it does, falling back to the default (which is right 99+% of the time) IMO is OK.
3) if it does fail, I would expect just about everything else to fail.

The other alternative would be to log to the event viewer, which we should still have access to here, if you're really concerned.

File: pkg/util/winutil/shutil.go (hunk 5 above)
Comment (truthbk) [reply pkg/util/winutil/shutil.go]: I think something needs reformatting in `pkg/util/winutil/shutil.go`, CI is failing.

File: cmd/agent/common/common_windows.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (truthbk) [reply cmd/agent/common/common_windows.go]: I think I would like to log something in the event of (2) - it's the kind of thing that might make us lose a ton of time debugging if we log nothing, whereas it'd provide a very useful datapoint if something strange happens here. I would prefer to log to our agent logs to be totally honest, but if that's problematic the event viewer might be better than nothing. Event viewer agent entries are not shipped with the flare, right?

File: cmd/agent/app/integrations_windows.go [hunk 7]
```diff
@@ -30,6 +31,8 @@ func init() {
 	pd, err := winutil.GetProgramDataDir()
 	if err == nil {
 		tufPipCachePath = filepath.Join(pd, "Datadog", "repositories", "cache")
+	} else {
+		winutil.LogEventViewer(config.ServiceName, 0x8000000F, tufPipCachePath)
```
Comment (truthbk) [inline cmd/agent/app/integrations_windows.go]: Not a huge fan of these magic numbers, but I can see we use them elsewhere. At some point we should maybe use consts and bit masks for the log level and message number. Can we create a card for that?

File: pkg/util/winutil/eventlog.go [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+func LogEventViewer(servicename string, msgnum uint32, arg string) {
+	elog, err := eventlog.Open(servicename)
+	if err != nil {
+		return
+	}
+	defer elog.Close()
+	switch msgnum & 0xF0000000 {
+	case 0x40000000:
+		// Info level message
+		elog.Info(msgnum, arg)
```
Comment (truthbk) [inline pkg/util/winutil/eventlog.go]: Just want to make sure...

is the `msgnum` we want to pass here the full message number, or do we have to mask out the most significant bits?

**Review by truthbk — COMMENTED** [review]
I think I'm satisfied with this now as far as I'm concerned. I'd like us to cleanup the magic number stuff, but in a different PR - no need to do it here.

Can we remove the `go dep` stuff and do that in a separate PR?

File: pkg/util/winutil/eventlog.go (hunk 8 above)
Comment (derekwbrown) [reply pkg/util/winutil/eventlog.go]: It's the full message number.  The high order 4 bits just indicates if it's info/warning/error level.  So the `case` is just so we send it to the right `log` method.  But, we need the whole message number.

**Review by truthbk — DISMISSED** [review]
Looks good to me! Thanks!
📦

**Pull request merged**